Unix/http/sessionmap.c

```diff
@@ -0,0 +1,318 @@
+#include "sessionmap.h"
+#include <pal/memory.h>
+#include <pal/strings.h>
+#include <assert.h>
+#include <common.h>
+
+#define ENABLE_TRACING 1
+#define FORCE_TRACING 1
+
+#ifdef ENABLE_TRACING
+# define TRACING_LEVEL 4
+#include <base/result.h>
+#include <base/logbase.h>
+#include <base/log.h>
+#define LOGE2 __LOGE
+#define LOGW2 __LOGW
+#define LOGD2 __LOGD
+#else
+# define LOGE2(a)
+# define LOGW2(a)
+# define LOGD2(a)
+# define LOGX2(a)
+#endif
+
+typedef struct _SessionBucket
+{
+    struct _SessionBucket *next;// HashMap usage
+    const char* sessionId;      // WS_MAN Session Id 
+    size_t hash;                // Hash value for sessionId
+    char* cookie;               // MS-WSMAN Cookie
+                                // allocated on creation and update.
+                                // owned by the bucket
+} SessionBucket;
+
+#define SESSIONID_HASH_SEED 1313038763
+
+static size_t _SessionId_Hash(_In_z_ const char* sessionId)
+{
+    size_t hash = SESSIONID_HASH_SEED;
+    if (sessionId)
+    {
+        const char*p = sessionId;
+        while(*p)
+        {
+            // From strset.c
+            hash ^= ((hash << 5) + *p + (hash >> 2));
+            p++;
+        }
+    }
+    return hash;
+}
+
+static size_t _SessionBucket_GetHash(_In_ const HashBucket *b)
+{
+    const SessionBucket *bucket = (const SessionBucket*) b;
+    return bucket->hash;
+}
+
+static int _SessionBucket_Equal(_In_ const HashBucket* b1, _In_ const HashBucket* b2)
+{
+    const SessionBucket* bucket1 = (const SessionBucket*) b1;
+    const SessionBucket* bucket2 = (const SessionBucket*) b2;
+    return 
+    (
+        bucket1->hash == bucket2->hash
+        &&
+        strcmp(bucket1->sessionId, bucket2->sessionId) == 0
+    );
+}
+
+static void _SessionBucket_Release(HashBucket *b)
+{
+    if (b)
+    {
+        SessionBucket* bucket = (SessionBucket*) b;
+        if (bucket->cookie)
+        {
+            PAL_Free(bucket->cookie);           
+        }
+        PAL_Free(bucket);
+    }
+}
+
+static SessionBucket* _SessionBucket_New(SessionMap *self, _In_z_ const char* sessionId, size_t hash)
+{
+    // WriteLock expected.
+    SessionBucket* bucket = NULL;
+    if (sessionId) do
```

Comment:
Personally I would prefer if and do on different lines, where do is nested inside curly brackets. It is cleaner to read.

Reply:
current style cop tools seem to recommend this pattern. Personally, I prefer it as well to avoid the extra indenting.    I'll add a // while(FALSE) comment for clarity unless you feel strongly enough about it.